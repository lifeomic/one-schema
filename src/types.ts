import type { JSONSchema4 } from 'json-schema';

export type EndpointDefinition = {
  Name: string;
  Request?: JSONSchema4;
  Response: JSONSchema4;
};

export type OneSchemaMetaDefinition = {
  PackageJSON?: Record<string, unknown>;
};

export type OneSchemaDefinition = {
  Meta?: OneSchemaMetaDefinition;

  Resources?: {
    [key: string]: JSONSchema4;
  };

  Endpoints: {
    [key: string]: EndpointDefinition;
  };
};

export type GeneratedEndpointsType = {
  [key: string]: {
    Request: any;
    PathParams: any;
    Response: any;
  };
};

export type OneSchema<Endpoints extends GeneratedEndpointsType> =
  OneSchemaDefinition & {
    Endpoints: {
      [K in keyof Endpoints]: {
        Request?: JSONSchema4;
        Response: JSONSchema4;
      };
    };
    __endpoints_type__?: Endpoints;
  };

export type EndpointsOf<Schema extends OneSchema<any>> = NonNullable<
  Schema['__endpoints_type__']
>;
