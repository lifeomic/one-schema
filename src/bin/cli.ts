#!/usr/bin/env node
import { mkdirSync, writeFileSync } from 'fs';
import * as path from 'path';
import yargs = require('yargs');
import { format, BuiltInParserName } from 'prettier';
import { dump } from 'js-yaml';

import { generateAxiosClient } from '../generate-axios-client';
import { generateAPITypes } from '../generate-api-types';
import { loadSchemaFromFile, SchemaAssumptions } from '../meta-schema';
import { toOpenAPISpec } from '../openapi';
import { fetchRemoteSchema } from '../fetch-remote-schema';

const getPrettierParser = (outputFilename: string): BuiltInParserName => {
  const extension = path.extname(outputFilename).replace('.', '');
  if (['yml', 'yaml'].includes(extension)) {
    return 'yaml';
  }
  if (extension === 'json') {
    return 'json';
  }
  return 'typescript';
};

const writeGeneratedFile = async (
  filepath: string,
  content: string,
  options: { format: boolean },
) => {
  mkdirSync(path.dirname(filepath), { recursive: true });
  writeFileSync(
    filepath,
    options.format
      ? await format(content, { parser: getPrettierParser(filepath) })
      : content,
    { encoding: 'utf-8' },
  );
};

const VALID_ASSUMPTION_KEYS: (keyof SchemaAssumptions)[] = [
  'noAdditionalPropertiesOnObjects',
  'objectPropertiesRequiredByDefault',
];

const parseAssumptions = (input: string): SchemaAssumptions => {
  const containsOnlyValidAssumptionKeys = (
    arr: string[],
  ): arr is (keyof SchemaAssumptions)[] =>
    arr.every((value) => VALID_ASSUMPTION_KEYS.includes(value as any));

  if (input === 'all') {
    return {
      noAdditionalPropertiesOnObjects: true,
      objectPropertiesRequiredByDefault: true,
    };
  }

  if (input === 'none') {
    return {
      noAdditionalPropertiesOnObjects: false,
      objectPropertiesRequiredByDefault: false,
    };
  }

  const assumptions = input.split(',');

  if (!containsOnlyValidAssumptionKeys(assumptions)) {
    throw new Error(
      "Detected an invalid assumptions input. Must be either 'all', 'none', or a comma-separated list containing one or more of: " +
        VALID_ASSUMPTION_KEYS.join(', '),
    );
  }

  return assumptions.reduce((accum, next) => ({ ...accum, [next]: true }), {
    requestsAreObjects: false,
    responsesAreObjects: false,
    noAdditionalPropertiesOnObjects: false,
    objectPropertiesRequiredByDefault: false,
  });
};

const getCommonOptions = (argv: yargs.Argv) =>
  argv
    .option('schema', {
      type: 'string',
      description: 'The filepath of the schema spec.',
      demandOption: true,
    })
    .option('output', {
      type: 'string',
      description: 'A filepath for the generated output.',
      demandOption: true,
    })
    .option('format', {
      type: 'boolean',
      description: 'Whether to format the output using prettier.',
      default: true,
    })
    .option('assumptions', {
      type: 'string',
      description:
        "Which JSONSchema assumptions to apply. Must be either 'all', 'none', or a comma-separated list containing one or more of: " +
        VALID_ASSUMPTION_KEYS.join(', '),
      default: 'all',
    });

export const createProgram = (argv: readonly string[]) =>
  yargs(argv)
    .command(
      'generate-axios-client',
      'Generates an Axios client using the specified schema and options.',
      (y) =>
        getCommonOptions(y).option('name', {
          type: 'string',
          description: 'The name of the generated client class.',
          demandOption: true,
        }),
      async (argv) => {
        const spec = loadSchemaFromFile(
          argv.schema,
          parseAssumptions(argv.assumptions),
        );
        const output = await generateAxiosClient({
          spec,
          outputClass: argv.name,
        });

        await writeGeneratedFile(argv.output, output.typescript, {
          format: argv.format,
        });
      },
    )
    .command(
      'generate-api-types',
      'Generates API types using the specified schema and options.',
      getCommonOptions,
      async (argv) => {
        const spec = loadSchemaFromFile(
          argv.schema,
          parseAssumptions(argv.assumptions),
        );

        const output = await generateAPITypes({ spec });

        await writeGeneratedFile(argv.output, output, { format: argv.format });
      },
    )
    .command(
      'generate-open-api-spec',
      'Generates an OpenAPI v3.1.0 spec using the specified schema and options.',
      (y) =>
        getCommonOptions(y)
          .option('apiVersion', {
            type: 'string',
            description: 'The current version of this API schema.',
            default: '1.0.0',
          })
          .option('apiTitle', {
            type: 'string',
            description: 'The API title.',
            demandOption: true,
          }),
      async (argv) => {
        const spec = loadSchemaFromFile(
          argv.schema,
          parseAssumptions(argv.assumptions),
        );

        const openAPISpec = toOpenAPISpec(spec, {
          info: { version: argv.apiVersion, title: argv.apiTitle },
        });

        const output =
          argv.output.endsWith('.yml') || argv.output.endsWith('.yaml')
            ? dump(openAPISpec, {
                /**
                 * Without this, js-yaml will default to a line width of 80 and use
                 * the "folded" multiline style. While this should not actually affect
                 * serialization or deserialization, it can result in ugly-looking output
                 * that contains newlines in unexpected places.
                 *
                 * This option allows us to preserve the original developer's newlines.
                 */
                lineWidth: -1,
              })
            : JSON.stringify(openAPISpec, null, 2);

        await writeGeneratedFile(argv.output, output, { format: argv.format });
      },
    )
    .command(
      'fetch-remote-schema',
      'Fetches a schema from a remote service via introspection.',
      (y) =>
        y
          .option('from', {
            type: 'string',
            description: 'The url of the remote schema.',
            demandOption: true,
          })
          .option('output', {
            type: 'string',
            description: 'A filepath for the fetched schema.',
            demandOption: true,
          }),
      async (argv) => {
        const result = await fetchRemoteSchema({ url: argv.from });

        await writeGeneratedFile(
          argv.output,
          JSON.stringify(
            {
              // Re-declaring here so that `serviceVersion` will end up at the top of the
              // file, and is clearly visible in source control.
              serviceVersion: result.serviceVersion,
              schema: result.schema,
            },
            null,
            2,
          ),
          { format: false },
        );
      },
    )
    .demandCommand()
    .strict()
    .exitProcess(false);

if (require.main === module) {
  createProgram(process.argv.slice(2))
    .parseAsync()
    .catch((err) => {
      console.error(err);
      process.exitCode = 1;
    });
}
