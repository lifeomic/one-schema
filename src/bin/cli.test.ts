import { describe, expect, it, test } from 'vitest';
import { readFileSync, writeFileSync } from 'fs';
import { tmpNameSync } from 'tmp';
import { OneSchemaDefinition } from '../types';
import * as yaml from 'js-yaml';
import { createProgram } from './cli';

/**
 * Writes the `schema`, and returns its filepath.
 */
const writeSchema = (schema: OneSchemaDefinition): string => {
  const path = `${__dirname}/../test-schema.json`;
  writeFileSync(path, JSON.stringify(schema, null, 2), { encoding: 'utf8' });
  return path;
};

describe('schema validation errors', () => {
  const SCENARIOS: {
    test: string;
    schema: string;
    expectOutputContaining: string;
  }[] = [
    {
      test: 'malformed yaml',
      schema: '-this is:bad',
      expectOutputContaining: 'Detected invalid schema: data must be object',
    },
  ];

  SCENARIOS.forEach(({ test: name, schema, expectOutputContaining }) => {
    test(`${name}`, async () => {
      const filepath = tmpNameSync();

      writeFileSync(filepath, schema.trim(), { encoding: 'utf-8' });

      await expect(
        createProgram([
          'generate-api-types',
          '--schema',
          filepath,
          '--output',
          tmpNameSync(),
        ]).parseAsync(),
      ).rejects.toThrow(expectOutputContaining);
    });
  });
});

describe('input validation errors', () => {
  const SCENARIOS: {
    test: string;
    input: string;
    expectOutputContaining: string;
  }[] = [
    {
      test: 'empty input',
      input: '',
      expectOutputContaining: 'Unknown argument: ""',
    },
    {
      test: 'bogus command name',
      input: 'generate-bogus',
      expectOutputContaining: 'Unknown argument: generate-bogus',
    },
    {
      test: 'missing arguments - generate-axios-client',
      input: 'generate-axios-client',
      expectOutputContaining:
        'Missing required arguments: schema, output, name',
    },
    {
      test: 'missing arguments - generate-api-types',
      input: 'generate-api-types',
      expectOutputContaining: 'Missing required arguments: schema, output',
    },
    {
      test: 'missing arguments - generate-open-api-spec',
      input: 'generate-open-api-spec',
      expectOutputContaining:
        'Missing required arguments: schema, output, apiTitle',
    },
    {
      test: 'missing arguments - fetch-remote-schema',
      input: 'fetch-remote-schema',
      expectOutputContaining: 'Missing required arguments: from, output',
    },
  ];

  SCENARIOS.forEach(({ test: name, input, expectOutputContaining }) => {
    test(`${name}`, () => {
      expect(() => createProgram([input]).parse()).toThrow(
        expectOutputContaining,
      );
    });
  });
});

describe('generate-open-api-spec', () => {
  it('does not create new newlines when serializing + deserializing to YAML', async () => {
    const description =
      'This is a long description that might go over the js-yaml default line length.\nIt also contains newlines.\n\nLots of newlines!';

    const path = writeSchema({
      Endpoints: {
        'GET /something': {
          Name: 'getSomething',
          Description: description,
          Response: {
            type: 'object',
          },
        },
      },
    });

    const output = `${__dirname}/../test-generated.yaml`;

    await createProgram([
      'generate-open-api-spec',
      '--schema',
      path,
      '--output',
      output,
      '--apiTitle',
      'test title',
    ]).parseAsync();

    const yamlContent = readFileSync(output, { encoding: 'utf8' });

    // Assert the output looks right.
    expect(yamlContent.trim()).toStrictEqual(
      `
openapi: 3.0.0
info:
  version: 1.0.0
  title: test title
components: {}
paths:
  /something:
    get:
      operationId: getSomething
      description: |-
        This is a long description that might go over the js-yaml default line length.
        It also contains newlines.

        Lots of newlines!
      responses:
        "200":
          description: A successful response
          content:
            application/json:
              schema:
                additionalProperties: false
                type: object
`.trim(),
    );

    // Assert the output deserializes correctly.
    const openapi: any = yaml.load(yamlContent);

    expect(openapi.paths['/something'].get.description).toStrictEqual(
      description,
    );
  });
});
