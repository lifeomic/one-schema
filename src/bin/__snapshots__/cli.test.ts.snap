// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`input validation snapshots bogus command name 1`] = `
"cli.ts <command>

Commands:
  cli.ts generate-axios-client   Generates an Axios client using the specified
                                 schema and options.
  cli.ts generate-api-types      Generates API types using the specified schema
                                 and options.
  cli.ts generate-open-api-spec  Generates an OpenAPI v3.1.0 spec using the
                                 specified schema and options.

Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]

Unknown argument: generate-bogus"
`;

exports[`input validation snapshots empty input 1`] = `
"cli.ts <command>

Commands:
  cli.ts generate-axios-client   Generates an Axios client using the specified
                                 schema and options.
  cli.ts generate-api-types      Generates API types using the specified schema
                                 and options.
  cli.ts generate-open-api-spec  Generates an OpenAPI v3.1.0 spec using the
                                 specified schema and options.

Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]

Not enough non-option arguments: got 0, need at least 1"
`;

exports[`input validation snapshots missing arguments - API generation 1`] = `
"cli.ts generate-api-types

Generates API types using the specified schema and options.

Options:
  --help         Show help                                             [boolean]
  --version      Show version number                                   [boolean]
  --schema       The filepath of the schema spec.            [string] [required]
  --output       A filepath for the generated output.        [string] [required]
  --format       Whether to format the output using prettier.
                                                      [boolean] [default: false]
  --assumptions  Which JSONSchema assumptions to apply. Must be either 'all',
                 'none', or a comma-separated list containing one or more of:
                 noAdditionalPropertiesOnObjects,
                 objectPropertiesRequiredByDefault    [string] [default: \\"none\\"]

Missing required arguments: schema, output"
`;

exports[`input validation snapshots missing arguments - Axios generation 1`] = `
"cli.ts generate-axios-client

Generates an Axios client using the specified schema and options.

Options:
  --help         Show help                                             [boolean]
  --version      Show version number                                   [boolean]
  --schema       The filepath of the schema spec.            [string] [required]
  --output       A filepath for the generated output.        [string] [required]
  --format       Whether to format the output using prettier.
                                                      [boolean] [default: false]
  --assumptions  Which JSONSchema assumptions to apply. Must be either 'all',
                 'none', or a comma-separated list containing one or more of:
                 noAdditionalPropertiesOnObjects,
                 objectPropertiesRequiredByDefault    [string] [default: \\"none\\"]
  --className    The name of the generated client class.
                                                    [string] [default: \\"Client\\"]

Missing required arguments: schema, output"
`;
